### ISSUE TITLE
moment loss from mindboggle

### ISSUE THREAD

**Issue by jpainam** [osamarais/moments_loss#4]
```python
from __future__ import division
import argparse
import torch
from torch import nn
import numpy as np
import scipy
from scipy.special import (factorial,
                        comb as nchoosek,
                        )

import logging
LOG = logging.getLogger(__name__)

#import decorator

#@decorator.decorator
#def logcall(fn, *args, **dargs):
#    LOG.debug(fn.__name__)
#    return fn(*args, **dargs)


# IMAG_CONST = scipy.sqrt(-1)
IMAG_CONST = np.lib.scimath.sqrt(-1)
PI_CONST = np.pi
NAN_CONST = np.NaN

from moment_helpers import autocat, nest
import timeit


class MomentLoss(nn.Module):
    def __init__(self, N):
        super(MomentLoss, self).__init__()
        self.N = N
        
    def forward(self, faces_pred, verts_pred, faces_gt, verts_gt):
        moments_pred = self._geometric_moments_exact(verts_pred, faces_pred, self.N)
        print(moments_pred.shape)
        moments_gt = self._geometric_moments_exact(verts_gt, faces_gt, self.N)
        print(moments_gt.shape)
        #print(moments_gt)
        #print(moments_pred)
        return torch.linalg.norm(torch.tensor(moments_pred) - torch.tensor(moments_gt))
    
    
    def _geometric_moments_exact(self, points_array, faces_array, N):
        n_facets, n_vertices = faces_array.shape[:2]
        assert n_vertices == 3
        moments_array = np.zeros([N + 1, N + 1, N + 1])
        monomial_array = self.monomial_precalc(points_array, N)
        for face in faces_array:
            vertex_list = [points_array[_i, ...] for _i in face]
            Cf_list = [monomial_array[_i, ...] for _i in face]
            Vf = self.facet_volume(vertex_list)
            moments_array += Vf * self.term_Sijk(Cf_list, N)
        return self.factorial_scalar(N) * moments_array

    def factorial_scalar(self, N):
        i, j, k = np.mgrid[0:N + 1, 0:N + 1, 0:N + 1]
        return factorial(i) * factorial(j) * factorial(k) / (factorial(i + j + k + 2) * (i + j + k + 3))

    def monomial_precalc(self, points_array, N):
        n_points = points_array.shape[0]
        monomial_array = np.zeros([n_points, N + 1, N + 1, N + 1])
        tri_array = self.trinomial_precalc(N)
        for point_indx, point in enumerate(points_array):
            monomial_array[point_indx, ...] = self.mon_comb(
                point, tri_array, N)
        return monomial_array

    def mon_comb(self, vertex, tri_array, N, out=None):
        x, y, z = vertex
        c = np.zeros([N + 1, N + 1, N + 1])
        for i, j, k in nest(lambda: range(N + 1),
                            lambda _i: range(N - _i + 1),
                            lambda _i, _j: range(N - _i - _j + 1),
                            ):
            c[i, j, k] = tri_array[i, j, k] * \
                np.power(x, i) * np.power(y, j) * np.power(z, k)
        return c

    def term_Sijk(self, Cf_list, N):
        S = np.zeros([N + 1, N + 1, N + 1])
        C0, C1, C2 = Cf_list
        Dabc = self.term_Dabc(C1, C2, N)
        for i, j, k, ii, jj, kk in nest(lambda: range(N + 1),
                                        lambda _i: range(N - _i + 1),
                                        lambda _i, _j: range(N - _i - _j + 1),
                                        lambda _i, _j, _k: range(_i + 1),
                                        lambda _i, _j, _k, _ii: range(_j + 1),
                                        lambda _i, _j, _k, _ii, _jj: range(
                                            _k + 1),
                                        ):
            S[i, j, k] += C0[ii, jj, kk] * Dabc[i - ii, j - jj, k - kk]
        return S

    def trinomial_precalc(self, N):
        tri_array = np.zeros([N + 1, N + 1, N + 1])
        for i, j, k in nest(lambda: range(N + 1),
                            lambda _i: range(N - _i + 1),
                            lambda _i, _j: range(N - _i - _j + 1)
                            ):
            tri_array[i, j, k] = self.trinomial(i, j, k)
        return tri_array

    def trinomial(self, i, j, k):
        return factorial(i + j + k) / (factorial(i) * factorial(j) * factorial(k))

    def facet_volume(self, vertex_list):
        return np.linalg.det(autocat(vertex_list, axis=1))

    def term_Dabc(self, C1, C2, N):
        D = np.zeros([N + 1, N + 1, N + 1])
        for i, j, k, ii, jj, kk in nest(lambda: range(N + 1),
                                        lambda _i: range(N + 1),
                                        lambda _i, _j: range(N + 1),
                                        lambda _i, _j, _k: range(_i + 1),
                                        lambda _i, _j, _k, _ii: range(_j + 1),
                                        lambda _i, _j, _k, _ii, _jj: range(
                                            _k + 1)
                                        ):
            D[i, j, k] += C1[ii, jj, kk] * C2[i - ii, j - jj, k - kk]
        return D




parser = argparse.ArgumentParser(
                    prog='Moment Loss',
                    description='Compute the moment loss',
                    epilog='Moment Loss')
parser.add_argument('-N', type=int, default=1)

if __name__ == "__main__":
    start = timeit.default_timer()
    
    args = parser.parse_args()
    
    moment_loss = MomentLoss(N=args.N)
    v_matrix = [[0, 0, 0], [0, 0, 1], [0, 1 ,0], [0, 1, 1], [1, 0, 0], [1, 0, 1], [1, 1 ,0], [1, 1 ,1]]
    f_matrix = [[0,2,4], [4,2,6],[0,1,3], [0,3,2], [0,5,1], [0,4,5], [1,7,3], [1,5,7], [4,7,5], [4,6,7], [7,6,2],[7,2,3]]

    faces_pred = np.array(f_matrix)
    faces_gt = np.array(f_matrix)
    
    verts_pred = np.array(v_matrix)
    verts_gt = np.array(v_matrix)
    loss = moment_loss(faces_pred, verts_pred, faces_gt, verts_gt)
    print('loss : %.4f' % loss)
    print('[{}] Completed'.format(timeit.default_timer() - start))



```

**Comment by jpainam** [comment]
add the helpers function
```python
import numpy as np

def nested_loop(stack, args):
    if len(stack) != 0:
        fn = stack.pop()
        for i in fn(*args):
            for j in nested_loop(stack, args+[i]):
                yield (i,)+j
        stack.append(fn)
    else:
        yield tuple()

def nest(*_stack):
    return nested_loop(list(reversed(_stack)), [])

def autocat(arrs, **dargs):
    axis = dargs.pop('axis', None)
    if axis is None:
        return np.concatenate(arrs, **dargs)
    ndim = arrs[0].ndim
    assert all([ a.ndim == ndim for a in arrs])
    if axis >= ndim:
        arrs = tuple([ np.expand_dims(a,axis) for a in arrs ])
    return np.concatenate(arrs, axis=axis)

def main():

    def zero(): return list(range(0, 3))
    def one(_x): return list(range(0, _x))
    def two(_x, _y): return list(range(0, _y))

    for i in nest(zero, one, two):
        print('{0} {1}'.format('#'*5, i))

if __name__ == '__main__':
    main()
```
